clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -1060,238 +1061,348 @@ CodeGenFunction::evaluateOrEmitBuiltinObjectSize(const Expr *E, unsigned Type,
+    if (Idx)
+      Cmp = Builder.CreateAnd(Builder.CreateIsNotNeg(Index), Cmp);
+    return Builder.CreateSelect(Cmp, Res,
+                                ConstantInt::get(ResType, 0, IsSigned));
+  };
+
+  if (FlexibleArrayMemberFD == FD) {
+    if (Idx) { // Option (2) '&ptr->array[idx]'
+      //  size_t index_size = index * flexible_array_member_base_size;
+      Value *IndexSize = Builder.CreateMul(FlexibleArrayMemberBaseSize, Index,
+                                           "index_size", !IsSigned, IsSigned);
+
+      //  return flexible_array_member_size - index_size;
+      return CheckForNegative(Builder.CreateSub(
+          FlexibleArrayMemberSize, IndexSize, "result", !IsSigned, IsSigned));
+    } else { // Option (1) 'ptr->array'
+      //  return flexible_array_member_size;
+      return CheckForNegative(FlexibleArrayMemberSize);
+    }
+  } else {
+    //  size_t sizeof_struct = sizeof (struct s);
+    llvm::StructType *StructTy = getTypes().getCGRecordLayout(RD).getLLVMType();
+    const llvm::DataLayout &Layout = CGM.getDataLayout();
+    TypeSize Size = Layout.getTypeSizeInBits(StructTy);
+    Value *SizeofStruct =
+        llvm::ConstantInt::get(ResType, Size.getKnownMinValue() / CharWidth);
+
+    //  size_t field_offset = offsetof (struct s, field);
+    int64_t Offset = 0;
+    GetFieldOffset(Ctx, RD, FD, Offset);
+    Value *FieldOffset =
+        llvm::ConstantInt::get(ResType, Offset / CharWidth, IsSigned);
+
+    if (Idx) { // Option (4) '&ptr->field_array[idx]'
+      //  size_t field_base_size = sizeof (*ptr->field_array);
+      const ArrayType *ArrayTy = Ctx.getAsArrayType(FieldTy);
+      CharUnits BaseSize = Ctx.getTypeSizeInChars(ArrayTy->getElementType());
+      auto *FieldBaseSize =
+          llvm::ConstantInt::get(ResType, BaseSize.getQuantity(), IsSigned);
+
+      //  field_offset += index * field_base_size;
+      Value *Mul = Builder.CreateMul(Index, FieldBaseSize, "field_offset",
+                                     !IsSigned, IsSigned);
+      FieldOffset = Builder.CreateAdd(FieldOffset, Mul);
+
+      //  size_t offset_diff = flexible_array_member_offset - field_offset;
+      Value *OffsetDiff = Builder.CreateSub(SizeofStruct, FieldOffset,
+                                            "offset_diff", !IsSigned, IsSigned);
+
+      //  return offset_diff + flexible_array_member_size;
+      return CheckForNegative(
+          Builder.CreateAdd(FlexibleArrayMemberSize, OffsetDiff, "result"));
+    } else { // Option (3) '&ptr->field'
+      //  size_t offset_diff = flexible_array_member_offset - field_offset;
+      Value *OffsetDiff = Builder.CreateSub(SizeofStruct, FieldOffset,
+                                            "offset_diff", !IsSigned, IsSigned);
+
+      //  return flexible_array_member_size + offset_diff;
+      return CheckForNegative(
+          Builder.CreateAdd(FlexibleArrayMemberSize, OffsetDiff, "result"));
+    }
+  }
+
+  llvm_unreachable("unable to handle expression");
```

Comment:
You could avoid this unreachable by changing the control flow from:

```python
if ...:
  return
else ...:
  return
unreachable
```
to:
```python
if ...:
  return
return
```
(sorry, end of control flow nits. /me ducks)